dependencies.go

```diff
@@ -27,19 +30,116 @@ type (
 	}
 )
 
+// dependencyCommitFail corrupts the first page of a transaction when it
+// is committed
+type dependencyCommitFail struct {
+	prodDependencies
+}
+
+func (dependencyCommitFail) disrupt(s string) bool {
+	if s == "CommitFail" {
+		return true
+	}
+	return false
+}
+
+// dependencyReleaseFail corrupts the first page of a transaction when it
+// is released
+type dependencyReleaseFail struct {
+	prodDependencies
+}
+
+func (dependencyReleaseFail) disrupt(s string) bool {
+	if s == "ReleaseFail" {
+		return true
+	}
+	return false
+}
+
 // prodDependencies is a passthrough to the standard library calls
 type prodDependencies struct{}
 
 func (prodDependencies) disrupt(string) bool { return false }
-
 func (prodDependencies) readFile(path string) ([]byte, error) {
 	return ioutil.ReadFile(path)
 }
-
 func (prodDependencies) openFile(path string, flag int, perm os.FileMode) (file, error) {
 	return os.OpenFile(path, flag, perm)
 }
-
 func (prodDependencies) create(path string) (file, error) {
 	return os.Create(path)
 }
+
+// faultyFile implements a file that simulates a faulty disk.
+type faultyFile struct {
+	file   *os.File
+	failed bool
+
+	// failDenominator determins how likely it is that a write will fail, defined
+	// as 1/failDenominator. Each write call increments failDenominator, and it
+	// starts at 2. This means that the more calls to WriteAt, the less likely
+	// the write is to fail.
+	failDenominator int
+}
+
+func (f *faultyFile) Read(p []byte) (int, error) {
+	return f.file.Read(p)
+}
+func (f *faultyFile) Write(p []byte) (int, error) {
+	fail := fastrand.Intn(f.failDenominator) == 0
+	f.failDenominator++
+	if fail || f.failDenominator >= 5000 {
```

Comment:
always fail after 5000 writes? should be documented

Reply:
yeah, because the file is supposed to fail at some point. It becomes less likely to fail the more you write, so much so that 1 in 100,000 times or something you'll see like 1 million writes that get through. It's not inifinite, but large enough that we need to add a clamp.